tests/base/hdf5_05.cc

```diff
@@ -0,0 +1,1463 @@
+
+        pcout << "IO mode " + dataset_name << "<" << type_name << ">"
+              << " (Write): " << dataset.get_io_mode() << std::endl;
+        pcout << "Local no collective cause " + dataset_name << "<" << type_name
+              << ">"
+              << " (Write): " << dataset.get_local_no_collective_cause()
+              << std::endl;
+        pcout << "Global no collective cause " + dataset_name << "<"
+              << type_name << ">"
+              << " (Write): " << dataset.get_global_no_collective_cause()
+              << std::endl;
+      }
+    else
+      {
+        dataset.template write_none<Number>();
+        pcout << "IO mode " + dataset_name << "<" << type_name << ">"
+              << " (Write): " << dataset.get_io_mode() << std::endl;
+        pcout << "Local no collective cause " + dataset_name << "<" << type_name
+              << ">"
+              << " (Write): " << dataset.get_local_no_collective_cause()
+              << std::endl;
+        pcout << "Global no collective cause " + dataset_name << "<"
+              << type_name << ">"
+              << " (Write): " << dataset.get_global_no_collective_cause()
+              << std::endl;
+      }
+  }
+
+  {
+    std::string dataset_name("dataset_6");
+
+    // This dataset is a 7x12 matrix
+    // Output generated by h5dump
+    //  (0,0): -0, 0, 1, -1, 2, 3, -2, 4, 5, -3, 6, 7,
+    //  (1,0): -4, 8, 9, -5, 10, 11, -6, 12, 13, -7, 14, 15,
+    //  (2,0): -8, 16, 17, -9, 18, 19, -10, 20, 21, -11, 22, 23,
+    //  (3,0): -12, 0, 0, -13, 0, 0, -14, 0, 0, -15, 0, 0,
+    //  (4,0): -16, 24, 25, -17, 26, 27, -18, 28, 29, -19, 30, 31,
+    //  (5,0): -20, 32, 33, -21, 34, 35, -22, 36, 37, -23, 38, 39,
+    //  (6,0): -24, 40, 41, -25, 42, 43, -26, 44, 45, -27, 46, 47
+
+    const std::vector<hsize_t> dataset_dimensions = {7, 12};
+    auto                       dataset =
+      group.create_dataset<Number>(dataset_name, dataset_dimensions);
+    dataset.set_query_io_mode(true);
+    deallog << "Dimensions " << dataset_name << "<" << type_name << ">"
+            << " (Write): " << dataset.get_dimensions() << std::endl;
+    deallog << "Size " << dataset_name << "<" << type_name << ">"
+            << " (Write): " << dataset.get_size() << std::endl;
+    deallog << "Rank " << dataset_name << "<" << type_name << ">"
+            << " (Write): " << dataset.get_rank() << std::endl;
+
+    if (Utilities::MPI::this_mpi_process(mpi_communicator) == 0)
+      {
+        const std::vector<hsize_t> data_dimensions = {6, 8};
+        auto data = initialize_container<FullMatrix, Number>(data_dimensions);
+        assign_data(data);
+        const std::vector<hsize_t> offset = {0, 1};
+        const std::vector<hsize_t> stride = {4, 3};
+        const std::vector<hsize_t> count  = {2, 4};
+        const std::vector<hsize_t> block  = {3, 2};
+
+        dataset.write_hyperslab(
+          data, data_dimensions, offset, stride, count, block);
```

Comment:
This writes a complex hyperslab. This might not be very useful in the context of deal.ii, but it is useful for other HPC applications
A complex hyperslab looks like this:
![kvx0p](https://user-images.githubusercontent.com/25910642/47932564-495b0380-ded2-11e8-8e40-47f4a038486f.png)